File: mobile/src/features/chat/verseActions.ts
import { Share } from "react-native";
import * as Clipboard from "expo-clipboard";
import type { GetToken } from "@/lib/api";
import type { RetrievedVerse } from "@/lib/chatView";
import type { TranslationId } from "@/features/bible/translations";
import { createNote, deleteNote, patchNote } from "@/features/notes/api";

type VerseActionSource = Pick<RetrievedVerse, "reference" | "text"> & {
	translation?: TranslationId;
};

/** Parse a route value without allowing an arbitrary string to change reader state. */
export function parseTranslationId(value: unknown): TranslationId | null {
	return value === "KJV" || value === "NKJV" ? value : null;
}

/** A source-card translation wins only for that reader route, never globally. */
export function readerTranslation(
	accountTranslation: TranslationId,
	sourceTranslation: unknown,
): TranslationId {
	return parseTranslationId(sourceTranslation) ?? accountTranslation;
}

/** Build the chapter route for a retrieved source while preserving legacy routes. */
export function readerRouteParams(
	target: { order: number; chapter: number; verse?: number },
	translation?: TranslationId,
): { book: string; chapter: string; verse?: string; translation?: TranslationId } {
	return {
		book: String(target.order),
		chapter: String(target.chapter),
		...(target.verse ? { verse: String(target.verse) } : {}),
		...(translation ? { translation } : {}),
	};
}

/** "John 3:16 — \"For God so loved...\" (KJV)" plain-text form for copy/share. */
export function formatVerseForSharing(
	verse: VerseActionSource,
	translation?: TranslationId
): string {
	const body = verse.text?.trim();
	const label = translation ?? verse.translation;
	const suffix = label ? ` (${label})` : "";
	return body ? `${verse.reference} — "${body}"${suffix}` : `${verse.reference}${suffix}`;
}

/** The narrowly-scoped source attribution carried by a Daily Cross CTA. */
export interface VerseAttachmentOrigin {
	surface: "daily-cross";
	verseOfDayId: string;
	reference: string;
	action: "go-deeper";
}

/** A verse or whole chapter the user attached to their next chat question. */
export interface VerseAttachment {
	reference: string;
	text: string;
	translation: TranslationId;
	origin?: VerseAttachmentOrigin;
}

/**
 * Compose the outgoing user message for /api/ask-question: the formatted
 * passage first, then the user's own question. No canned prompt — when the
 * question is empty the passage goes out on its own.
 */
export function composeMessageWithAttachment(
	question: string,
	attachment: VerseAttachment | null
): string {
	const trimmed = question.trim();
	if (!attachment) return trimmed;
	const verseBlock = formatVerseForSharing(attachment, attachment.translation);
	return trimmed ? `${verseBlock}\n\n${trimmed}` : verseBlock;
}

export async function copyVerse(verse: VerseActionSource, translation?: TranslationId): Promise<void> {
	await Clipboard.setStringAsync(formatVerseForSharing(verse, translation));
}

export async function shareVerse(verse: VerseActionSource, translation?: TranslationId): Promise<void> {
	await Share.share({ message: formatVerseForSharing(verse, translation) });
}

/**
 * Save a verse to the notes library: creates a note titled by the reference
 * with the passage as a Scripture blockquote (HTML, so it round-trips with the
 * web/mobile rich text editors). Returns the new note id.
 */
export async function saveVerseToNote(
	getToken: GetToken,
	verse: VerseActionSource,
	translation?: TranslationId
): Promise<string> {
	const text = verse.text?.trim() ?? "";
	const label = translation ?? verse.translation;
	const translationParagraph = label ? `<p>(${escapeHtml(label)})</p>` : "";
	const htmlContent =
		`<blockquote><p><strong>${escapeHtml(verse.reference)}</strong></p>` +
		(text ? `<p>${escapeHtml(text)}</p>` : "") +
		translationParagraph +
		"</blockquote>";
	const plainText = formatVerseForSharing(verse, translation);
	const wordCount = plainText.split(/\s+/).filter(Boolean).length;

	const note = await createNote(getToken, { title: verse.reference, folderId: null });
	try {
		await patchNote(getToken, note.id, { htmlContent, plainText, wordCount });
	} catch (error) {
		// Best-effort cleanup: without the content PATCH the note is an empty
		// orphan titled by the reference, so remove it before reporting failure.
		try {
			await deleteNote(getToken, note.id);
		} catch {
			// Keep the original failure.
		}
		throw error;
	}
	return note.id;
}

function escapeHtml(value: string): string {
	return value
		.replace(/&/g, "&amp;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;")
		.replace(/"/g, "&quot;");
}

File: macos/Shared/Chat/VerseActions.swift
#if os(macOS)
import AppKit
#else
import UIKit
#endif
import Foundation

/// A verse or whole chapter the user attached to their next question.
/// Port of `mobile/src/features/chat/verseActions.ts`.
struct VerseAttachment: Sendable, Equatable {
    var reference: String
    var text: String
    var translation: TranslationID
}

extension VerseAttachment {
    /// `John 3:16 — "For God so loved…" (KJV)` — the plain-text form used for
    /// copy, share, and the attachment block prepended to a question.
    static func formatForSharing(
        reference: String,
        text: String?,
        translation: TranslationID = .kjv
    ) -> String {
        let body = text?.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let body, !body.isEmpty else { return "\(reference) (\(translation.rawValue))" }
        return "\(reference) — \"\(body)\" (\(translation.rawValue))"
    }

    /// Compose the outgoing user message: the passage first, then the question.
    /// There is no canned prompt — an empty question sends the passage alone.
    static func compose(_ question: String, attachment: VerseAttachment?) -> String {
        let trimmed = question.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let attachment else { return trimmed }
        let block = formatForSharing(
            reference: attachment.reference,
            text: attachment.text,
            translation: attachment.translation
        )
        return trimmed.isEmpty ? block : "\(block)\n\n\(trimmed)"
    }

    var summary: String {
        Self.formatForSharing(reference: reference, text: text, translation: translation)
    }
}

// MARK: - Verse actions

enum VerseActions {
    static func copy(reference: String, text: String?, translation: TranslationID = .kjv) {
        let value = VerseAttachment.formatForSharing(
            reference: reference,
            text: text,
            translation: translation
        )
        #if os(macOS)
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(value, forType: .string)
        #else
        UIPasteboard.general.string = value
        #endif
    }

    /// Save a verse as its own note, titled by the reference with the passage as
    /// a Scripture blockquote. The HTML shape matches the other two clients so
    /// the note round-trips through their rich-text editors.
    @discardableResult
    static func saveToNote(
        api: APIClient,
        reference: String,
        text: String?,
        translation: TranslationID = .kjv
    ) async throws -> String {
        let body = text?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let html = "<blockquote><p><strong>\(escapeHTML(reference))</strong></p>"
            + (body.isEmpty ? "" : "<p>\(escapeHTML(body))</p>")
            + "<p>(\(escapeHTML(translation.rawValue)))</p>"
            + "</blockquote>"
        let plainText = VerseAttachment.formatForSharing(
            reference: reference,
            text: text,
            translation: translation
        )
        let wordCount = plainText.split(whereSeparator: \.isWhitespace).count

        struct NewNote: Encodable {
            let title: String
            let folderId: String?
        }
        struct NoteContent: Encodable {
            let htmlContent: String
            let plainText: String
            let wordCount: Int
        }
        struct NoteRef: Decodable { let id: String }

        let note = try await api.json(
            "/api/notes",
            method: "POST",
            body: NewNote(title: reference, folderId: nil),
            as: NoteRef.self
        )
        do {
            try await api.data(
                "/api/notes/\(note.id)",
                method: "PATCH",
                body: NoteContent(htmlContent: html, plainText: plainText, wordCount: wordCount)
            )
        } catch {
            // Best-effort cleanup: without the content PATCH the note is an
            // empty orphan titled by the reference, so remove it before
            // reporting failure.
            try? await api.data("/api/notes/\(note.id)", method: "DELETE")
            throw error
        }
        return note.id
    }

    private static func escapeHTML(_ value: String) -> String {
        value
            .replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
            .replacingOccurrences(of: "\"", with: "&quot;")
    }
}
